contracts/LibSubmarineSimple.sol

```diff
@@ -0,0 +1,195 @@
+    /////////////
+
+    /*
+       Keeping these functions makes instantiating a contract more expensive for gas costs, but helps with testing
+    */
+
+    function getCommitId(
+        address _user,
+        address _libsubmarine,
+        uint256 _commitValue,
+        bytes _embeddedDAppData,
+        bytes32 _witness,
+        uint256 _gasPrice,
+        uint256 _gasLimit
+    ) public pure returns (bytes32) {
+        return keccak256(abi.encodePacked(_user, _libsubmarine, _commitValue, _embeddedDAppData, _witness, _gasPrice, _gasLimit));
+    }
+    
+    function getCommitState(bytes32 _commitId) public view returns (
+        bool unlocked,
+        bool revealed,
+        uint256 amountRevealed,
+        uint256 amountUnlocked
+    ) {
+        CommitData memory sesh = commitData[_commitId];
+        return (
+            sesh.revealed,
+            sesh.unlocked,
+            sesh.amountRevealed,
+            sesh.amountUnlocked
+        );
+    }
+
+    /////////////
+    // Setters //
+    /////////////
+
+    /**
+     * @notice Function called by the user to reveal the session.
+     * @dev warning Must be called within 256 blocks of the commit transaction to obtain the correct blockhash.
+     * @param _commitBlockNumber Block number in which the commit tx was included.
+     * @param _unlockAmount Value (i.e. how much money was sent) included in the commit tx.
+     * @param _embeddedDAppData optional Data passed embedded within the unlock tx. This should probably be null
+     * @param _witness Witness "secret" we committed to
+     * @param _unlockGasPrice Gas price of the unlock TX
+     * @param _unlockGasLimit Gas limit of the unlock TX
+     * @param _unlockTXHash full hash of the signed unlock transaction
+     * @param _proofBlob the Proof blob that gets passed to ethprove to verify merkle tree inclusion in a prior block.
+     */
+    function reveal(
+        uint32 _commitBlockNumber,
+        bytes _embeddedDAppData,
+        uint256 _unlockAmount,
+        bytes32 _witness,
+        uint256 _unlockGasPrice,
+        uint256 _unlockGasLimit,
+        bytes32 _unlockTXHash,
+        bytes _proofBlob
+    ) public {
+        bytes32 commitBlockHash = blockhash(_commitBlockNumber);
+        // fullCommit = (addressA + addressC + aux(sendAmount) + dappData + w + aux(gasPrice) + aux(gasLimit))
+
+        bytes32 commitId = getCommitId(
+            msg.sender,
+            address(this),
+            _unlockAmount,
+            _embeddedDAppData,
+            _witness,
+            _unlockGasPrice,
+            _unlockGasLimit
+        );
+
+        require(commitData[commitId].revealed == false, "The tx is already revealed");
+        require(commitBlockHash != 0x0, "Commit Block is too old to retreive block hash (more than 256 blocks), or does not exist");
+        require(block.number - _commitBlockNumber > commitPeriodLength, "You must wait long enough to allow the committing period to end before revealing");
```

Comment:
What is the reason for the commit period?

Reply:
Also, you can remove the safemath libraries since they are no longer used.

Reply:
The commit period is the core aspect that prevents frontrunning. The contract won't accept reveals until after the commit period is over - so this forces that gap in blocks that's necessary to ensure that a commit is locked into the chain long before the reveal reveals the existence of a commitment and gives away all the information a evil party needs to do to perform frontrunning

Reply:
Ah okay, an other way approach is `require(block.number != _commitBlockNumber)`. The additional blocks given increased security against reordering?

Reply:
yeah, the other approach works for an auction contract, but there are some contexts (e.g. dex contracts) where there may not be hardcodable block number or the desired end to the commit period may not be known at contract constructor-time

Security against reordering, but more importantly, frontrunning - forces the commit to have happened in the past, otherwise there's no strong requirement for the commit to have happened in the past.